### PR TITLE
feat: setting instance name during component creation

### PR DESCRIPTION
**Pull request by umutsoysalansys**
## Description
**This change enables users to set the instance name (optional) while creating a component. **
image

## Issue linked
**Please mention the issue number or describe the problem this pull request addresses.**

## Checklist
- [x] I have tested my changes locally.
- [x] I have added necessary documentation or updated existing documentation.
- [x] I have followed the coding style guidelines of this project.
- [x] I have added appropriate unit tests.
- [x] I have reviewed my changes before submitting this pull request.
- [x] I have linked the issue or issues that are solved to the PR if any.
- [x] I have assigned this PR to myself.
- [x] I have added the minimum version decorator to any new backend method implemented.
- [x] I have made sure that the title of my PR follows [Conventional commits style](https://www.conventionalcommits.org/en/v1.0.0/#summary) (e.g. ``feat: extrude circle to cylinder``)

### CODE REVIEW

**Comment by umutsoysalansys** [comment]
@b-matteo @RobPasMue ready for review:)

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [inline src/ansys/geometry/core/designer/component.py:416]:
```suggestion
                        template=template,
                        instance_name=instance_name,
```
Any optional argument, please pass it with its identifier...

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [inline src/ansys/geometry/core/designer/component.py:198]: You should also initialize to None the instance name here
```suggestion
                self._id = None
                self._instance_name = None
```

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [inline src/ansys/geometry/core/designer/component.py:195]:
```suggestion
                self._instance_name = new_component.component.instance_name
```
You should get this value from the response of the creation.

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [inline src/ansys/geometry/core/designer/component.py:178]:
```suggestion
            self._instance_name = instance_name
```
Typo...

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [inline src/ansys/geometry/core/designer/component.py:180]:
```suggestion
```
@umutsoysalansys - this is unacceptable. Make sure you never commit print statements in a PR. It's fine if you want to debug with them locally but you should get used to using the debugger.

File: tests/integration/test_design.py
Comment (RobPasMue) [inline tests/integration/test_design.py:385]:
```suggestion
    assert nested_component.instance_name is None
```
You should check using the property... and not directly accessing the private attribute. In some occasions you just "have to" but not here since you have access to the property...

File: tests/integration/test_design.py
Comment (RobPasMue) [inline tests/integration/test_design.py:386]:
```suggestion
    assert nested_component2.instance_name == "first instance"
```

Comment (RobPasMue) [reply tests/integration/test_design.py:386]: Same here

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [inline src/ansys/geometry/core/designer/component.py:163]: This argument is not documented. Please document it.

**Review by RobPasMue — CHANGES_REQUESTED** [review]
Several cleanup comments and improvements. Also, @umutsoysalansys - have you checked with @rajesh1359 or @agvarghe that this is what they were expecting? I'm not sure where the request for this is coming from

**Comment by RobPasMue** [comment]
Please update your branch with respect to main as well.

File: src/ansys/geometry/core/designer/component.py
Comment (umutsoysalansys) [reply src/ansys/geometry/core/designer/component.py:195]: I did it this way for a reason. In API server we create the component but then return an occurence of which doesn't share the instance name. @b-matteo

File: src/ansys/geometry/core/designer/component.py
Comment (umutsoysalansys) [reply src/ansys/geometry/core/designer/component.py:180]: oh thank you roberto, mea culpa.

**Comment by umutsoysalansys** [comment]
@RobPasMue, yes this resolves their issue of duplicated instance name which is currently happening in main branch.  Here is an example script that you can reproduce to see the issue. 
`
from ansys.geometry.core import launch_modeler_with_spaceclaim
from ansys.geometry.core.math import UNITVECTOR3D_X, Point2D
from ansys.geometry.core.misc import UNITS, Distance
from ansys.geometry.core.sketch import Sketch
from ansys.geometry.core import *

modeler = launch_modeler_with_spaceclaim()
#modeler = Modeler(host="localhost", port= 50051)

design = modeler.create_design("DesignHierarchyExample")
 
circle_sketch = Sketch()
circle_sketch.circle(Point2D([10, 10], UNITS.mm), Distance(10, UNITS.mm))
 
cylinder_body = design.extrude_sketch("10mmCylinder", circle_sketch, Distance(10, UNITS.mm))
 
slot_sketch = Sketch()
slot_sketch.slot(Point2D([40, 10], UNITS.mm), Distance(20, UNITS.mm), Distance(10, UNITS.mm))
 
nested_component = design.add_component("NestedComponent")
slot_body = nested_component.extrude_sketch("SlotExtrusion", slot_sketch, Distance(20, UNITS.mm))
 
design.plot()

if __name__ == "__main__":
    pass
`

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [reply src/ansys/geometry/core/designer/component.py:196]: I still think we should use the response value. The occurrence returned should contain that instance name IMO.

**Comment by RobPasMue** [comment]
> @RobPasMue, yes this resolves their issue of duplicated instance name which is currently happening in main branch. Here is an example script that you can reproduce to see the issue.

I know it solves it but have you verified with them that this is the expected usage on their side or were they expecting a different API usage/call?

File: src/ansys/geometry/core/designer/component.py
Comment (umutsoysalansys) [reply src/ansys/geometry/core/designer/component.py:196]: I think we can merge it this way because it is time sensitive and then we should have a talk with Matteo to make sure it is not against the intended occurrence design there.

**Comment by umutsoysalansys** [comment]
Yes I did, no they don't.

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [reply src/ansys/geometry/core/designer/component.py:196]: I get it is time sensitive but we should do things properly and not rush things. Otherwise we end up like in #1396 - which we haven't solved in weeks already.

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [reply src/ansys/geometry/core/designer/component.py:199]: Do not resolve comments you have not committed @umutsoysalansys

File: src/ansys/geometry/core/designer/component.py
Comment (umutsoysalansys) [reply src/ansys/geometry/core/designer/component.py:196]: @b-matteo, @RobPasMue. I'm done with this PR, I'll let you guys decide what you want to do. Feel free to update it. Thanks.

**Review by RobPasMue — CHANGES_REQUESTED** [review]
If you are resolving comments, make sure you are either answering them or committing my suggestions. Do not just close them @umutsoysalansys

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [reply src/ansys/geometry/core/designer/component.py:196]: @umutsoysalansys - a PR is not done until it is merged.

File: src/ansys/geometry/core/designer/component.py
Comment (umutsoysalansys) [reply src/ansys/geometry/core/designer/component.py:199]: sorry I can't follow you really.

File: src/ansys/geometry/core/designer/component.py
Comment (RobPasMue) [inline src/ansys/geometry/core/designer/component.py:388]:
```suggestion
        self, name: str, template: Optional["Component"] = None, instance_name: str = None
```

**Comment by RobPasMue** [comment]
Windows new version is working. Linux version is having issues. @b-matteo investigating.

File: tests/integration/test_design.py
Comment (umutsoysalansys) [inline tests/integration/test_design.py]: @RobPasMue why this?

File: src/ansys/geometry/core/designer/component.py
Comment (umutsoysalansys) [inline src/ansys/geometry/core/designer/component.py:199]: @smereu

File: tests/integration/test_design.py
Comment (umutsoysalansys) [inline tests/integration/test_design.py:385]: kind of unnecessary since we already have other tests cover this. but whatever @RobPasMue

Comment (RobPasMue) [reply tests/integration/test_design.py:385]: When discussing with @b-matteo, we saw that name and instance_name were being overwritten in the server side by whatever was passed as instance_name. It would be interesting to see that the message returned from the service is consistent with what we expect and that no overriding of the name takes place. That is why we are actively checking as well both name and instance_name.

Tests are not expected to be "performant" but make the code reliable. The more things we can check (that have been demonstrated to cause an impact), the better.

**Comment by umutsoysalansys** [comment]
@RobPasMue, updated API Server, linux tests are passing. What is this testing with minimum requirement workflow? @b-matteo How to test locally?

**Comment by umutsoysalansys** [comment]
@RobPasMue pls review and approve thanks.

**Comment by RobPasMue** [comment]
> @RobPasMue, updated API Server, linux tests are passing. What is this testing with minimum requirement workflow? @b-matteo How to test locally?

Minimum requirement workflow just tries to install the project without the "test-pinned" dependencies. It runs on all supported Python versions, but it also uses the ``linux-latest`` image. Since we have not pushed the unstable ones to stable yet, the tests related to the instance_name. Don't worry @umutsoysalansys - once we push the images to stable, the checks will pass.

**Comment by umutsoysalansys** [comment]
[like]  Umut Soysal reacted to your message:
________________________________

**Comment by RobPasMue** [comment]
> @RobPasMue pls review and approve thanks.

You are not the only one wanting to finish this PR @umutsoysalansys - we are all waiting for things to work properly and pass in the workflow. Until then, we cannot merge. You can assume it's approved anyway.

**Comment by umutsoysalansys** [comment]
> > @RobPasMue pls review and approve thanks.
> 
> You are not the only one wanting to finish this PR @umutsoysalansys - we are all waiting for things to work properly and pass in the workflow. Until then, we cannot merge. You can assume it's approved anyway.

what is holding it back? Unstable images are passing? @RobPasMue

**Comment by RobPasMue** [comment]
> what is holding it back? Unstable images are passing? @RobPasMue

Linux latest unstable image is updated. ADO is still pushing the Windows image.

**Comment by umutsoysalansys** [comment]
> > what is holding it back? Unstable images are passing? @RobPasMue
> 
> Linux latest unstable image is updated. ADO is still pushing the Windows image.

uploded now, it should be about minutes to finish checks.

**Comment by RobPasMue** [comment]
@umutsoysalansys please stop, we have already heard you

**Review by RobPasMue — APPROVED** [review]
LGTM - merging

**Pull request merged**
